### PR TITLE
Resolve instantiation problem with init_meta_context

### LINKED ISSUE #10323: `init_meta_context` fails `isinstance` call (fails xFormers)
## 🐛 Bug

The below case fails for `init_meta_context`. This probably has to do something with the subclasses for `MyModule` being changed. This as a result fails when we use an [xformers](https://github.com/facebookresearch/xformers) model factory within the context manager. 

```python
import torch

from pytorch_lightning.utilities.meta import init_meta_context


class BaseModule(torch.nn.Module):
    pass


class MyModule(BaseModule):
    pass


with init_meta_context():
    my_module = MyModule()
    assert isinstance(my_module, BaseModule)
```

cc @tchaton @blefaudeux

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Closes #10323 

This PR adds support for properly using init_meta_context with more complex model like the ones used within xformer: https://github.com/facebookresearch/xformers

This doesn't work with DDPSpawn right now.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1404,10 +1404,23 @@ def _call_setup_hook(self) -> None:
...
             self.call_hook("configure_sharded_model")
             self.call_hook("on_configure_sharded_model")
 
+    def _handle_meta_model(self) -> None:
+        param = next(self.lightning_module.parameters())
+        if param.device.type != "meta":
+            return
+
+        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+            raise MisconfigurationException("LightningModule on meta device isn't supported with spawn.")
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1417]: does it work with TPUSpawn?

Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py:1417]: It is a subclass of DDPSpawn, so it should fail too

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1417]: No, it doesn't. Updated !

File: pytorch_lightning/utilities/meta.py [hunk 2]
```diff
@@ -312,12 +312,20 @@ def search(mod: ModuleType) -> List[ModuleType]:
             setattr(mod, subclass.__name__, _MetaClass)
 
 
+def mock_isinstance(A, B, isinstance=None):
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/meta.py:316]: Could you add a brief docstring why we need this?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1404,10 +1411,23 @@ def _call_setup_hook(self) -> None:
...
+            self._handle_meta_model()
             self.call_hook("configure_sharded_model")
             self.call_hook("on_configure_sharded_model")
 
+    def _handle_meta_model(self) -> None:
+        param = next(self.lightning_module.parameters())
+        if param.device.type != "meta":
+            return
+
+        if isinstance(self.training_type_plugin, (DDPSpawnPlugin, TPUSpawnPlugin)):
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/trainer.py:1423]:
```suggestion
        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
```
Not required, as TPUSpawn plugin is a subclass of it

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py]: oh no! the update isn't required 😅 since `TPUSpawn` subclasses `DDPSpawn`. I asked to make sure to avoid this in-case it works with `TPUSpawn`.

File: pytorch_lightning/utilities/meta.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/meta.py]: Done !

File: pytorch_lightning/core/mixins/device_dtype_mixin.py [hunk 4]
```diff
@@ -177,7 +179,9 @@ def __update_properties(
         self, device: Optional[torch.device] = None, dtype: Optional[Union[str, torch.dtype]] = None
     ) -> None:
         def apply_fn(module: Union["DeviceDtypeModuleMixin", Module]) -> None:
-            if not isinstance(module, DeviceDtypeModuleMixin):
+            # TODO: Find why `isinstance(module, DeviceDtypeModuleMixin)` doesn't
+            # work when using `init_meta_device`.
```
Comment (carmocca) [inline pytorch_lightning/core/mixins/device_dtype_mixin.py:183]: Isn't it worth trying to resolve this TODO here? might be because we have a bug

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1404,10 +1404,23 @@ def _call_setup_hook(self) -> None:
...
+        if param.device.type != "meta":
+            return
+
+        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+            raise MisconfigurationException("LightningModule on meta device isn't supported with spawn.")
+
+        materialize_module(self.lightning_module)
+        self.lightning_module.trainer = proxy(self)
+        # TODO: Find a better place to move the newly materialized model to the device
+        self.training_type_plugin.model_to_device()
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/trainer.py:1422]:
```suggestion
```

File: pytorch_lightning/core/mixins/device_dtype_mixin.py [hunk 6]
```diff
@@ -177,7 +179,9 @@ def __update_properties(
...
+            if not isinstance(module, (DeviceDtypeModuleMixin, pl.LightningModule)):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/core/mixins/device_dtype_mixin.py:184]: Done !

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -1404,10 +1404,23 @@ def _call_setup_hook(self) -> None:
...
     def _call_configure_sharded_model(self) -> None:
         with self.accelerator.model_sharded_context():
-            materialize_module(self.lightning_module)
+            self._handle_meta_model()
             self.call_hook("configure_sharded_model")
             self.call_hook("on_configure_sharded_model")
 
+    def _handle_meta_model(self) -> None:
+        param = next(self.lightning_module.parameters())
+        if param.device.type != "meta":
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1413]: Unless creating a util for this, no there is not.

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -1404,10 +1404,23 @@ def _call_setup_hook(self) -> None:
...
+            return
+
+        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+            raise MisconfigurationException("LightningModule on meta device isn't supported with spawn.")
+
+        materialize_module(self.lightning_module)
+        self.lightning_module.trainer = proxy(self)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1420]: Yes, the trainer reference is lost during the materialization.

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1422]: Removed !

File: pytorch_lightning/utilities/meta.py [hunk 9]
```diff
@@ -312,12 +314,22 @@ def search(mod: ModuleType) -> List[ModuleType]:
...
 @contextmanager
 def init_meta_context() -> Generator:
     rank_zero_warn(
         "Be aware this feature is highly experimental and there are a number of weird edge cases "
         "where it can internal assert and/or crash. A more stable version is to be expected from PyTorch 1.11."
     )
     _set_meta_device()
+    __builtins__["isinstance"] = partial(_mock_isinstance, isinstance=isinstance)
     yield
+    __builtins__["isinstance"] = isinstance.keywords["isinstance"]
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/meta.py:334]: I tried and it didn't work. Feel free to give it a try.

File: pytorch_lightning/trainer/trainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1420]:
```suggestion
        # the trainer reference is lost during materialization
        self.lightning_module.trainer = proxy(self)
```

File: pytorch_lightning/core/mixins/device_dtype_mixin.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/core/mixins/device_dtype_mixin.py]: The training seems to work fine without it. I have no idea why it doesn't work.

File: pytorch_lightning/trainer/trainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1421]: Done !

File: pytorch_lightning/utilities/meta.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/meta.py:334]: Without this trick, it won't work.

[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/utilities/meta.py:334]: I also gave this a try in a previous attempt to fix the issue, my skillz were not good enough!

**Review by SeanNaren — APPROVED** [review]
@tchaton you're a wizard! Definitely some cleanup may be warranted in the future, but we both tested that this worked with an example.

File: pytorch_lightning/utilities/meta.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/meta.py:334]: @carmocca Any luck ?

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Cleaned out !

File: pytorch_lightning/utilities/meta.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/utilities/meta.py]: Done with a2a1084fea96231f0a311ed2f6bc130dc376af70

Please give it a try, the test passes

**Review by carmocca — APPROVED** [review]
Top 3 weirdest code in our codebase :joy:

**Review by rohitgr7 — APPROVED** [review]
nit

File: CHANGELOG.md [hunk 10]
```diff
@@ -131,6 +131,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `CombinedLoader` and `max_size_cycle` didn't receive a `DistributedSampler` ([#10374](https://github.com/PyTorchLightning/pytorch-lightning/issues/10374))
 
 
+- Fixed `isinstance` not working with `init_meta_context`, materialize model not being moved to the device ([#10493](https://github.com/PyTorchLightning/metrics/pull/10493))
```
Comment (awaelchli) [inline CHANGELOG.md:134]:
```suggestion
- Fixed `isinstance` not working with `init_meta_context`, materialized model not being moved to the device ([#10493](https://github.com/PyTorchLightning/metrics/pull/10493))
```
?

File: pytorch_lightning/core/mixins/device_dtype_mixin.py [hunk 11]
```diff
@@ -177,7 +179,9 @@ def __update_properties(
         self, device: Optional[torch.device] = None, dtype: Optional[Union[str, torch.dtype]] = None
     ) -> None:
         def apply_fn(module: Union["DeviceDtypeModuleMixin", Module]) -> None:
-            if not isinstance(module, DeviceDtypeModuleMixin):
+            # TODO: Find why `isinstance(module, DeviceDtypeModuleMixin)` doesn't
```
Comment (awaelchli) [inline pytorch_lightning/core/mixins/device_dtype_mixin.py:182]: is the changelog outdated?

File: pytorch_lightning/trainer/trainer.py [hunk 12]
```diff
@@ -1406,10 +1406,21 @@ def _call_setup_hook(self) -> None:
 
     def _call_configure_sharded_model(self) -> None:
         with self.accelerator.model_sharded_context():
-            materialize_module(self.lightning_module)
+            self._handle_meta_model()
             self.call_hook("configure_sharded_model")
             self.call_hook("on_configure_sharded_model")
 
+    def _handle_meta_model(self) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1413]: why is this a new method? can't it be done in the method above?

File: pytorch_lightning/utilities/meta.py [hunk 13]
```diff
@@ -321,3 +327,11 @@ def init_meta_context() -> Generator:
     _set_meta_device()
     yield
     _unset_meta_device()
+
+
+def is_meta_device(module: nn.Module) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/meta.py:332]:
```suggestion
def on_meta_device(module: nn.Module) -> bool:
```
or
```suggestion
def is_on_meta_device(module: nn.Module) -> bool:
```
?

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md (hunk 10 above)
Comment (tchaton) [reply CHANGELOG.md:134]: Done !

File: pytorch_lightning/core/mixins/device_dtype_mixin.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/core/mixins/device_dtype_mixin.py:182]: This is the `materialized model not being moved to the device`. Somehow, the isinstance(module, DeviceDtypeModuleMixin) doesn't work with the materialised model.

File: pytorch_lightning/trainer/trainer.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1413]: I believe it is cleaner within its own method.

File: pytorch_lightning/utilities/meta.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/utilities/meta.py:332]: Changed for `is_on_meta_device`
